### PR TITLE
Use ruff instead of flake8

### PR DESCRIPTION
**Pull request by gaborbernat**
- Fix broken issue link and bump tools
- Use ruff instead of flake8

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -1,13 +1,16 @@
-"""Sphinx configuration file for pytest-memray documentation."""
+"""Sphinx configuration file for pytest-memray documentation."""  # noqa: INP001
```
Comment (godlygeek) [inline docs/conf.py:1]: Mm. It doesn't have any provision for handling scripts that aren't part of packages? That's annoying...

File: docs/conf.py [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
 import sys
 from pathlib import Path
 from subprocess import check_output
+from typing import TYPE_CHECKING
 
-from sphinx.application import Sphinx
 from sphinxcontrib.programoutput import Command
 
+if TYPE_CHECKING:
+    from sphinx.application import Sphinx
```
Comment (godlygeek) [inline docs/conf.py:12]: Why does this need to be moved below an `if TYPE_CHECKING:`?

File: docs/conf.py [hunk 3]
```diff
@@ -47,11 +50,11 @@ def _get_output(self):
...
 
 
-def setup(app: Sphinx) -> None:
+def setup(app: Sphinx) -> None:  # noqa: ARG001
     here = Path(__file__).parent
     root, exe = here.parent, Path(sys.executable)
     towncrier = exe.with_name(f"towncrier{exe.suffix}")
     cmd = [str(towncrier), "build", "--draft", "--version", "NEXT"]
-    new = check_output(cmd, cwd=root, text=True)
+    new = check_output(cmd, cwd=root, text=True)  # noqa: S603
```
Comment (godlygeek) [inline docs/conf.py:58]: We should probably ignore this rule globally. I have no idea what it's complaining about.

File: docs/demo/test_ok.py [hunk 4]
```diff
@@ -1,18 +1,18 @@
-from __future__ import annotations
+from __future__ import annotations  # noqa: INP001
```
Comment (godlygeek) [inline docs/demo/test_ok.py:1]: If this is in `docs/`, it's intended for end-user consumption, right? Not just for maintainers? If so, I don't think we should add annotations here to appease our linter. I think we should tell the linter to ignore this file instead.

File: pyproject.toml [hunk 5]
```diff
@@ -1,61 +1,65 @@
...
+  "hatch-vcs>=0.3",
+  "hatchling>=1.18",
+]
 
 [project]
 name = "pytest-memray"
 description = "A simple plugin to use with pytest"
-readme.file = "README.md"
 readme.content-type = "text/markdown"
+readme.file = "README.md"
```
Comment (godlygeek) [inline pyproject.toml:12]: Why was this line moved? Did a tool do this automatically?

File: pyproject.toml [hunk 6]
```diff
@@ -1,61 +1,65 @@
...
   "Programming Language :: Python :: 3.11",
   "Programming Language :: Python :: Implementation :: CPython",
   "Topic :: Software Development :: Debuggers",
 ]
-
+dynamic = [
+  "version",
+]
+dependencies = [
+  "memray>=1.8",
```
Comment (godlygeek) [inline pyproject.toml:37]: Why are we changing from depending on memray >=1.5 to >=1.8? Is this something that should have already been applied in the past and was missed? If so, can you at least make it a separate commit so we can find it when searching through the git log?

File: pyproject.toml [hunk 7]
```diff
@@ -1,61 +1,65 @@
...
+  "pytest>=7.3.2",
```
Comment (godlygeek) [inline pyproject.toml:38]: Likewise, why are we requiring a newer pytest version? Your commit message doesn't mention this change...

**Review by godlygeek — CHANGES_REQUESTED** [review]
As far as I can tell, you've got a mix of manual changes and automatically generated changes, as well as functional changes and stylistic changes, all rolled into a single commit. This is really hard to review. Can you please do separate commits for each logical change?

**Comment by gaborbernat** [comment]
> As far as I can tell, you've got a mix of manual changes and automatically generated changes, as well as functional changes and stylistic changes, all rolled into a single commit.

Nope, all was done automatically; minus a very few that ruff required to pass.

File: pyproject.toml (hunk 6 above)
Comment (gaborbernat) [reply pyproject.toml:37]: Because the CI is only testing the latest version of memray. Anything else is a guess at best, and a straight lie otherwise.

File: pyproject.toml (hunk 7 above)
Comment (gaborbernat) [reply pyproject.toml:38]: See my point from above.

File: pyproject.toml (hunk 5 above)
Comment (gaborbernat) [reply pyproject.toml:12]: Indeed used https://github.com/tox-dev/pyproject-fmt

File: docs/conf.py (hunk 3 above)
Comment (gaborbernat) [reply docs/conf.py:58]: I think locally is better to be explicit that this instance is allowed but not in general. https://beta.ruff.rs/docs/rules/#flake8-bandit-s shows the rules with explanation.

File: docs/conf.py (hunk 2 above)
Comment (gaborbernat) [reply docs/conf.py:12]: Done automatically by ruff, read https://beta.ruff.rs/docs/rules/typing-only-first-party-import/

File: docs/conf.py (hunk 1 above)
Comment (gaborbernat) [reply docs/conf.py:1]: All python files should be formatted, I don't see why this should be different.

File: docs/demo/test_ok.py (hunk 4 above)
Comment (gaborbernat) [reply docs/demo/test_ok.py:1]: I don't agree. Why should not we lint documentation files same as the rest?

File: docs/conf.py (hunk 1 above)
Comment (godlygeek) [reply docs/conf.py:1]: I think you misunderstood me. I'm not arguing that `docs/conf.py` shouldn't be formatted, I'm complaining that `ruff` thinks that `docs` is a Python package, and wants is to create a `docs/__init__.py` - it doesn't understand that `docs/conf.py` is a standalone script that's not part of any package.

File: docs/conf.py (hunk 3 above)
Comment (godlygeek) [reply docs/conf.py:58]: Yes, I read the rule explanation. It seems to complain about every call to `subprocess`, as far as I can see. I tested running it with `cmd` set to a list of hardcoded strings, and it still raised an S603 error. That doesn't seem useful to me - have I missed something?

File: docs/demo/test_ok.py (hunk 4 above)
Comment (godlygeek) [reply docs/demo/test_ok.py:1]: Because it makes the code harder to read by adding annotations that most of our users won't recognize or care about, which makes it function worse as documentation

File: pyproject.toml (hunk 6 above)
Comment (godlygeek) [reply pyproject.toml:37]: OK, fine. This should be done in a separate commit, so that your reviewer doesn't need to guess what changes were made automatically by a tool and what changes were made manually by you.

**Comment by godlygeek** [comment]
It's very, very difficult to review this when all in one commit lines have been moved and changed, with some changes having been done manually and others having been done by automated tools. I'd like at least 4 separate commits here:

1. The configuration changes to start using `ruff` instead of `flake8` and `isort`
2. The changes automatically made by `ruff check --fix`
3. The changes automatically made by `pyproject-fmt`
4. The changes manually made to the versions of our declared dependencies

Plus separate commits for any other manual changes I didn't notice.

**Comment by gaborbernat** [comment]
Let's forget it, too much effort.

**Comment by gaborbernat** [comment]
Thanks for the review though 🙏
